[mlir][spirv] Add gpu printf op lowering to spirv.CL.printf op

drprajap opened the pull request:
This change contains following:
	- adds lowering of printf op to spirv.CL.printf op in GPUToSPIRV pass.
	- Fixes Constant decoration parsing for spirv GlobalVariable.
	- minor modification to spirv.CL.printf op assembly format.

drprajap committed:
[mlir][spirv] Add gpu printf op lowering to spirv.CL.printf op

This change contains following:
	- adds lowering of printf op to spirv.CL.printf op in GPUToSPIRV pass.
	- Fixes Constant decoration parsing for spirv GlobalVariable.
	- minor modification to spirv.CL.printf op assembly format.

drprajap committed:
clang format fix

drprajap committed:
Fix test

kuhar commented on mlir/lib/Conversion/GPUToSPIRV/GPUToSPIRV.cpp:
```diff
@@ -607,6 +616,108 @@ class GPUSubgroupReduceConversion final
+  mlir::SmallString<16> globalVarName;
+  mlir::spirv::GlobalVariableOp globalVar;
+
+  // formulate spirv global variable name
+  do {
+    globalVarName.clear();
+    (formatStringPrefix + llvm::Twine(stringNumber++))
+        .toStringRef(globalVarName);
+  } while (moduleOp.lookupSymbol(globalVarName));
```
I don't understand what this is trying to do. Can we simplify this and/or add some comment with an example?

kuhar commented on mlir/lib/Conversion/GPUToSPIRV/GPUToSPIRV.cpp:
```diff
@@ -607,6 +616,108 @@ class GPUSubgroupReduceConversion final
+        .toStringRef(globalVarName);
+  } while (moduleOp.lookupSymbol(globalVarName));
+
+  auto i8Type = rewriter.getI8Type();
+  auto i32Type = rewriter.getI32Type();
```
```suggestion
  IntegerType i8Type = rewriter.getI8Type();
  IntegerType i32Type = rewriter.getI32Type();
```

kuhar commented on mlir/lib/Conversion/GPUToSPIRV/GPUToSPIRV.cpp:
```diff
@@ -607,6 +616,108 @@ class GPUSubgroupReduceConversion final
+  unsigned scNum = 0;
+  auto createSpecConstant = [&](unsigned value) {
+    auto attr = rewriter.getI8IntegerAttr(value);
+    mlir::SmallString<16> specCstName;
+    (llvm::Twine(globalVarName) + "_sc" + llvm::Twine(scNum++))
+        .toStringRef(specCstName);
```
same as above

kuhar commented on mlir/lib/Conversion/GPUToSPIRV/GPUToSPIRV.cpp:
```diff
@@ -607,6 +616,108 @@ class GPUSubgroupReduceConversion final
+  auto i8Type = rewriter.getI8Type();
+  auto i32Type = rewriter.getI32Type();
+
+  unsigned scNum = 0;
```
```suggestion
  unsigned specConstantNum = 0;
```

kuhar commented on mlir/lib/Conversion/GPUToSPIRV/GPUToSPIRV.cpp:
```diff
@@ -607,6 +616,108 @@ class GPUSubgroupReduceConversion final
+  mlir::SmallString<16> globalVarName;
+  mlir::spirv::GlobalVariableOp globalVar;
+
+  // formulate spirv global variable name
```
```suggestion
  // Formulate SPIR-V global variable name.
```

kuhar commented on mlir/lib/Conversion/GPUToSPIRV/GPUToSPIRV.cpp:
```diff
@@ -607,6 +616,108 @@ class GPUSubgroupReduceConversion final
+        loc, rewriter.getStringAttr(specCstName), attr);
+  };
+
+  // define GlobalVarOp with printf format string using SpecConstants
+  // and make composite of SpecConstants
```
```suggestion
  // Define a GlobalVarOp with printf format string using SpecConstants
  // and make composite of SpecConstants.
```

kuhar commented on mlir/lib/Conversion/GPUToSPIRV/GPUToSPIRV.cpp:
```diff
@@ -607,6 +616,108 @@ class GPUSubgroupReduceConversion final
+  // define GlobalVarOp with printf format string using SpecConstants
+  // and make composite of SpecConstants
+  {
+    mlir::Operation *parent =
+        mlir::SymbolTable::getNearestSymbolTable(gpuPrintfOp->getParentOp());
+
+    mlir::ConversionPatternRewriter::InsertionGuard guard(rewriter);
```
nit: we don't need the mlir namespace prefix

kuhar commented on mlir/lib/Conversion/GPUToSPIRV/GPUToSPIRV.cpp:
```diff
@@ -607,6 +616,108 @@ class GPUSubgroupReduceConversion final
+
+    // Create Constituents with SpecConstant to construct
+    // SpecConstantCompositeOp
+    llvm::SmallString<20> formatString(gpuPrintfOp.getFormat());
```
Shouldn't this grab the format from the adaptor instead?

kuhar commented on mlir/lib/Conversion/GPUToSPIRV/GPUToSPIRV.cpp:
```diff
@@ -607,6 +616,108 @@ class GPUSubgroupReduceConversion final
+  printfArgs.reserve(argsRange.size() + 1);
+  printfArgs.append(argsRange.begin(), argsRange.end());
+
+  rewriter.create<mlir::spirv::CLPrintfOp>(loc, i32Type, fmtStr, printfArgs);
+
+  rewriter.eraseOp(gpuPrintfOp);
```
Use `rewriter.replaceWithNewOp`.

drprajap replied:
I tried to use that met with a crash with below error
op->getNumResults() == newValues.size() && "incorrect # of replacement values"' failed
gpu.printf op doesn't take return type so I am not sure that creates an issue when using replaceOpWithNewOp.

kuhar commented on mlir/test/Conversion/GPUToSPIRV/printf.mlir:
```diff
@@ -0,0 +1,71 @@
+        gpu.return
+    }
+  }
+}
```
nit: missing newline

kuhar reviewed (changes requested):
Thanks for the contribution, look like a useful debugging feature to have! I left some comments.

I find it hard to follow the `matchAndRewrite` code -- could we brake it down a bit and introduce some helper functions?

MrSidims commented on mlir/lib/Conversion/GPUToSPIRV/GPUToSPIRV.cpp:
```diff
@@ -607,6 +616,108 @@ class GPUSubgroupReduceConversion final
+  mlir::Value fmtStr = rewriter.create<mlir::spirv::BitcastOp>(
+      loc,
+      mlir::spirv::PointerType::get(i8Type,
+                                    mlir::spirv::StorageClass::UniformConstant),
```
Should we instead check GV's storage class and use it instead of constant? Both SPIR-V and OpenCL have extensions relaxing printf string arg requirement, see https://github.com/KhronosGroup/SPIRV-Registry/pull/148

drprajap replied:
I am not sure I got your suggestion, do you mean we can get the storage class from already defined globalVar op? 

thanks for pointing out to the requirements docs, while implementing this, when I checked  OpenCL specs for printf , they mention fmt string needs to be in constant address space, hence I used UniformConstant here- https://registry.khronos.org/OpenCL/specs/3.0-unified/html/OpenCL_C.html#printf

drprajap committed:
address feedback

drprajap commented on mlir/include/mlir/Dialect/SPIRV/IR/SPIRVCLOps.td:
````diff
@@ -875,7 +875,7 @@ def SPIRV_CLPrintfOp : SPIRV_CLOp<"printf", 184, []> {
 
     ```mlir
-    %0 = spirv.CL.printf %0 %1 %2 : (!spirv.ptr<i8, UniformConstant>, (i32, i32)) -> i32
+    %0 = spirv.CL.printf %0 : !spirv.ptr<i8, UniformConstant>(%1, %2 : i32, i32) -> i32
````
Thanks for suggestion. I changed it now.

drprajap commented on mlir/lib/Conversion/GPUToSPIRV/GPUToSPIRV.cpp:
```diff
@@ -607,6 +616,108 @@ class GPUSubgroupReduceConversion final
+  auto funcOp =
+      rewriter.getBlock()->getParent()->getParentOfType<mlir::spirv::FuncOp>();
+
+  auto moduleOp = funcOp->getParentOfType<mlir::spirv::ModuleOp>();
```
I added a check to see if module is valid then only proceed, let me know if it looks okay. This lowering is intended to be used with spirv.func op.

drprajap commented:
> Thanks for the contribution, look like a useful debugging feature to have! I left some comments.
> 
> I find it hard to follow the `matchAndRewrite` code -- could we brake it down a bit and introduce some helper functions?

Thank you @kuhar for your feedback and sorry for the delay in addressing them. 
I added more comments and tried to clarify the functionality, let me know if it makes it a bit more clear.

kuhar commented on mlir/include/mlir/Dialect/SPIRV/IR/SPIRVCLOps.td:
```diff
@@ -875,7 +875,9 @@ def SPIRV_CLPrintfOp : SPIRV_CLOp<"printf", 184, []> {
-    %0 = spirv.CL.printf %0 %1 %2 : (!spirv.ptr<i8, UniformConstant>, (i32, i32)) -> i32
+    %0 = spirv.CL.printf %fmt %1, %2  : !spirv.ptr<i8, UniformConstant>, i32, i32 -> i32
+    
+   $format `,` ( $arguments^ )? attr-dict `:`  type($format) ( `,` type($arguments)^ )? `->` type($result)
```
The assembly format shouldn't be in the examples section

kuhar commented on mlir/lib/Conversion/GPUToSPIRV/GPUToSPIRV.cpp:
```diff
@@ -597,6 +606,120 @@ class GPUSubgroupReduceConversion final
+
+  Location loc = gpuPrintfOp.getLoc();
+
+  auto moduleOp = gpuPrintfOp->getParentOfType<spirv::ModuleOp>();
+
+  if (!moduleOp) {
+    return success();
+  }
```
`success` indicates that you matched, so we need to return `failure` here. Also some stylistic nits:
```suggestion
  auto moduleOp = gpuPrintfOp.getParentOfType<spirv::ModuleOp>();
  if (!moduleOp)
    return failure();
```

kuhar commented on mlir/lib/Conversion/GPUToSPIRV/GPUToSPIRV.cpp:
```diff
@@ -597,6 +606,120 @@ class GPUSubgroupReduceConversion final
+    return success();
+  }
+
+  const char formatStringPrefix[] = "printfMsg";
```
```suggestion
  StringRef formatStringPrefix = "printfMsg";
```

kuhar commented on mlir/lib/Conversion/GPUToSPIRV/GPUToSPIRV.cpp:
```diff
@@ -597,6 +606,120 @@ class GPUSubgroupReduceConversion final
+
+  const char formatStringPrefix[] = "printfMsg";
+  unsigned stringNumber = 0;
+  SmallString<16> globalVarName;
+  spirv::GlobalVariableOp globalVar;
+
+  // SPIR-V global variable is used to initialize printf
+  // format string value, if there are multiple printf messages,
+  // each global var needs to be created with a unique name.
+  // like printfMsg0, printfMsg1, ...
+  // Formulate unique global variable name after
+  // searching in the module for existing global variable names.
+  // This is to avoid name collision with existing global variables.
+  do {
+    globalVarName.clear();
+    (formatStringPrefix + llvm::Twine(stringNumber++))
+        .toStringRef(globalVarName);
+  } while (moduleOp.lookupSymbol(globalVarName));
```
Could you make this into a helper function? This one is already a bit lengthy. Also, I'd use plain `std::string` here as the small size is less than what I'd expect the regular SSO length to be (20-ish characters). You can use `Twine::str()` to produce it.

drprajap replied:
Done, moved this to helper function that is used to construct globalVar and SpecConstant. It is using std::string and Twine::str() now.

kuhar commented on mlir/lib/Conversion/GPUToSPIRV/GPUToSPIRV.cpp:
```diff
@@ -597,6 +606,120 @@ class GPUSubgroupReduceConversion final
+  // @printfMsg0_sc0, @printfMsg0_sc1, ... by searching in the module
+  // for existing spec constant names.
+  unsigned specConstantNum = 0;
+  auto createSpecConstant = [&](unsigned value) {
+    auto attr = rewriter.getI8IntegerAttr(value);
+    SmallString<16> specCstName;
+    (llvm::Twine(globalVarName) + "_sc" + llvm::Twine(specConstantNum++))
+        .toStringRef(specCstName);
+
+    return rewriter.create<spirv::SpecConstantOp>(
+        loc, rewriter.getStringAttr(specCstName), attr);
+  };
```
Similar here: could we outline this code to a helper function?

drprajap replied:
Done.

kuhar commented on mlir/lib/Conversion/GPUToSPIRV/GPUToSPIRV.cpp:
```diff
@@ -597,6 +606,120 @@ class GPUSubgroupReduceConversion final
+
+  // Get printf arguments
+  auto argsRange = adaptor.getArgs();
+  SmallVector<Value, 4> printfArgs;
+  printfArgs.reserve(argsRange.size() + 1);
+  printfArgs.append(argsRange.begin(), argsRange.end());
```
Use `llvm::to_vector` if you need to

kuhar commented on mlir/lib/Conversion/GPUToSPIRV/GPUToSPIRV.cpp:
```diff
@@ -597,6 +606,120 @@ class GPUSubgroupReduceConversion final
+  printfArgs.reserve(argsRange.size() + 1);
+  printfArgs.append(argsRange.begin(), argsRange.end());
+
+  rewriter.create<spirv::CLPrintfOp>(loc, i32Type, fmtStr, printfArgs);
+
+  rewriter.eraseOp(gpuPrintfOp);
```
Could you add a comment that we need this because the spirv op produces a result that we need to discard?

kuhar commented:
@drprajap there are some test failures: https://buildkite.com/llvm-project/github-pull-requests/builds/101724#019202b6-5c05-43e1-aa01-48aba3a6a4de/6-202

kuhar commented:
@drprajap could you click the 'Resolve' button on past comments that you believe have been addressed? This will make it easier for me to tell what are the remaining open discussion threads.

drprajap committed:
Update mlir/lib/Conversion/GPUToSPIRV/GPUToSPIRV.cpp

review feedback

Co-authored-by: Jakub Kuderski <kubakuderski@gmail.com>

drprajap committed:
feedback - generalize varname lookup and minor changes

drprajap commented:
> @drprajap there are some test failures: https://buildkite.com/llvm-project/github-pull-requests/builds/101724#019202b6-5c05-43e1-aa01-48aba3a6a4de/6-202

Yes, sorry I missed to update test cases with new printf format changes, updated now.

drprajap commented:
> @drprajap could you click the 'Resolve' button on past comments that you believe have been addressed? This will make it easier for me to tell what are the remaining open discussion threads.

@kuhar I addressed all the feedback and resolved comments, please check it out and let me know if looks okay. Thanks so much for your valuable feedback.

kuhar commented on mlir/lib/Conversion/GPUToSPIRV/GPUToSPIRV.cpp:
```diff
@@ -597,6 +606,124 @@ class GPUSubgroupReduceConversion final
+/// searching in the module for existing variable/constant names.
+/// This is to avoid name collision with existing variables.
+/// Example: printfMsg0, printfMsg1, printfMsg2, ...
+std::string constructVarName(spirv::ModuleOp moduleOp, llvm::StringRef prefix) {
```
```suggestion
static std::string makeVarName(spirv::ModuleOp moduleOp, llvm::StringRef prefix) {
```

kuhar commented on mlir/lib/Conversion/GPUToSPIRV/GPUToSPIRV.cpp:
```diff
@@ -597,6 +606,124 @@ class GPUSubgroupReduceConversion final
   }
 };
 
+/// Formulate unique variable/constant name after
```
```suggestion
/// Formulate a unique variable/constant name after
```

kuhar commented on mlir/lib/Conversion/GPUToSPIRV/GPUToSPIRV.cpp:
```diff
@@ -597,6 +606,124 @@ class GPUSubgroupReduceConversion final
+    // Create Constituents with SpecConstant by scanning format string
+    // Each character of format string is stored as a spec constant
+    // and then these spec constants are used to create a
+    // SpecConstantCompositeOp
```
```suggestion
    // SpecConstantCompositeOp.
```

kuhar commented on mlir/lib/Conversion/GPUToSPIRV/GPUToSPIRV.cpp:
```diff
@@ -597,6 +606,124 @@ class GPUSubgroupReduceConversion final
+    // and then these spec constants are used to create a
+    // SpecConstantCompositeOp
+    llvm::SmallString<20> formatString(adaptor.getFormat());
+    formatString.push_back('\0'); // Null terminate for C
```
```suggestion
    formatString.push_back('\0'); // Null terminate for C.
```

kuhar commented on mlir/lib/Conversion/GPUToSPIRV/GPUToSPIRV.cpp:
```diff
@@ -597,6 +606,124 @@ class GPUSubgroupReduceConversion final
+    llvm::SmallString<20> formatString(adaptor.getFormat());
+    formatString.push_back('\0'); // Null terminate for C
+    SmallVector<Attribute, 4> constituents;
+    for (auto c : formatString) {
```
```suggestion
    for (char c : formatString) {
```

kuhar commented on mlir/lib/Conversion/GPUToSPIRV/GPUToSPIRV.cpp:
```diff
@@ -597,6 +606,124 @@ class GPUSubgroupReduceConversion final
+    formatString.push_back('\0'); // Null terminate for C
+    SmallVector<Attribute, 4> constituents;
+    for (auto c : formatString) {
+      auto cSpecConstantOp = createSpecConstant(c);
```
I wouldn't use `auto` here, the type is not obvious to me based on the RHS only.

kuhar commented on mlir/lib/Conversion/GPUToSPIRV/GPUToSPIRV.cpp:
```diff
@@ -597,6 +606,124 @@ class GPUSubgroupReduceConversion final
+  auto createSpecConstant = [&](unsigned value) {
+    auto attr = rewriter.getI8IntegerAttr(value);
+    std::string specCstName =
+        constructVarName(moduleOp, (llvm::Twine(globalVarName) + "_sc").str());
```
You can make `constructVarName` take `Twine` as the prefix string.

kuhar commented on mlir/lib/Conversion/GPUToSPIRV/GPUToSPIRV.cpp:
```diff
@@ -597,6 +606,124 @@ class GPUSubgroupReduceConversion final
+      globalPtr);
+
+  // Get printf arguments
+  SmallVector<Value, 4> printfArgs = llvm::to_vector<4>(adaptor.getArgs());
```
```suggestion
  auto printfArgs = llvm::to_vector_of<Value, 4>(adaptor.getArgs());
```

kuhar reviewed (commented):
Overall looks good, just some remaining issues

drprajap committed:
Apply suggestions from code review

formatting changes

Co-authored-by: Jakub Kuderski <kubakuderski@gmail.com>

drprajap committed:
function rename and use Twine as func arg

drprajap commented:
> Overall looks good, just some remaining issues

Thank you. Addressed remaining issues.

kuhar commented on mlir/lib/Conversion/GPUToSPIRV/GPUToSPIRV.cpp:
```diff
@@ -597,6 +606,124 @@ class GPUSubgroupReduceConversion final
+  auto createSpecConstant = [&](unsigned value) {
+    auto attr = rewriter.getI8IntegerAttr(value);
+    std::string specCstName =
+        makeVarName(moduleOp, (llvm::Twine(globalVarName) + "_sc"));
```
```suggestion
        makeVarName(moduleOp, llvm::Twine(globalVarName) + "_sc");
```

kuhar commented on mlir/lib/Conversion/GPUToSPIRV/GPUToSPIRV.cpp:
```diff
@@ -597,6 +606,124 @@ class GPUSubgroupReduceConversion final
+      spirv::PointerType::get(i8Type, spirv::StorageClass::UniformConstant),
+      globalPtr);
+
+  // Get printf arguments
```
```suggestion
  // Get printf arguments.
```

kuhar reviewed (approved):
LGTM

kuhar commented on mlir/test/Conversion/GPUToSPIRV/printf.mlir:
```diff
@@ -0,0 +1,71 @@
+// RUN: mlir-opt -allow-unregistered-dialect -split-input-file  -convert-gpu-to-spirv -verify-diagnostics %s -o - | FileCheck %s
```
Do we need the `-o o`?

drprajap replied:
Not really, was just trying to follow test command lines from other tests, I can remove it.

drprajap commented on mlir/test/Conversion/GPUToSPIRV/printf.mlir:
```diff
@@ -0,0 +1,71 @@
+// RUN: mlir-opt -allow-unregistered-dialect -split-input-file  -convert-gpu-to-spirv -verify-diagnostics %s -o - | FileCheck %s
```
```suggestion
// RUN: mlir-opt -allow-unregistered-dialect -split-input-file  -convert-gpu-to-spirv -verify-diagnostics %s | FileCheck %s
```

drprajap committed:
Apply suggestions from code review

minor formatting changes

Co-authored-by: Jakub Kuderski <kubakuderski@gmail.com>

drprajap commented:
> LGTM

Thanks, I do not have write permission, can you please help merge it if possible? else I can ask other team members.

kuhar commented:
> > LGTM
> 
> Thanks, I do not have write permission, can you please help merge it if possible? else I can ask other team members.

Sure, I will merge this on Monday. Please ping me if I forget to do so.

drprajap commented:
> > > LGTM
> > 
> > 
> > Thanks, I do not have write permission, can you please help merge it if possible? else I can ask other team members.
> 
> Sure, I will merge this on Monday. Please ping me if I forget to do so.

Gentle ping to merge this. Thank you :)
